### PR TITLE
CMS-211 Code review

### PR DESCRIPTION
**Pull request by sdonk**

### CODE REVIEW

File: config/settings.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/1.11/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = 'doj6z+3ryj)gn$v88vw03u)(z7*989%e)9c&p^p$r=(qu$797g'
```
Comment (sdonk) [inline config/settings.py]: This shouldn't be here. Move it to an env var

File: config/settings.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = True
```
Comment (sdonk) [inline config/settings.py]: Env var, default to `False`

File: config/settings.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
+    },
+]
+
+
+# Internationalization
+# https://docs.djangoproject.com/en/1.11/topics/i18n/
+
+LANGUAGE_CODE = 'en-us'
```
Comment (sdonk) [inline config/settings.py]: `en-gb`

File: config/settings.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+# Static files (CSS, JavaScript, Images)
+# https://docs.djangoproject.com/en/1.11/howto/static-files/
+
+STATIC_ROOT = os.path.join(BASE_DIR, 'staticfiles')
+STATIC_HOST = os.environ.get('STATIC_HOST', '')
+STATIC_URL = STATIC_HOST + '/static/'
+STATICFILES_STORAGE = 'whitenoise.storage.CompressedManifestStaticFilesStorage'
+
+
+PIR_API_KEY = os.getenv('PIR_API_KEY', 'changeme')
```
Comment (sdonk) [inline config/settings.py]: This should have no default, we want it not to run if this is not set

File: pir_frontend/admin.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
```
Comment (sdonk) [inline pir_frontend/admin.py]: remove this file if unused

File: pir_frontend/tests.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
```
Comment (sdonk) [inline pir_frontend/tests.py]: add tests

File: pir_frontend/views.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    context = {}
+
+    if request.method == 'POST':
+        form = PIRForm(request.POST)
+        if form.is_valid():
+            data = form.cleaned_data
+
+            try:
+                form.client.create_report(data)
+            except Exception as e:
```
Comment (sdonk) [inline pir_frontend/views.py]: Catching all exception is not good practice. The client should raise its own exception and that's what we need to catch here.

File: requirements.txt [hunk 8]
```diff
@@ -0,0 +1,9 @@
+Django==1.11.13
+pytz==2018.4
+whitenoise==3.3.1
+gunicorn==19.7.1
+django-countries==5.2
+django-nocaptcha-recaptcha==0.0.20
+directory-components==2.4.0
+export-elements==0.12.2
+git+https://github.com/uktrade/pir-api-client.git@0.0.6#egg=pir_client
```
Comment (sdonk) [inline requirements.txt]: this needs to be in pypi

**Comment by sdonk** [comment]
This repo is lacking all the backbone structure our repos have. 

1. It should have been called `pir-ui` for consistency
2. We need docker files, circle ci, makefile. Follow structure from here https://github.com/uktrade/invest-ui
3. This needs sentry to be configured

File: config/settings.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/1.11/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = os.environ['SECRET_KEY']
+
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = True if (os.getenv('DEBUG') == 'true') else False
```
Comment (richtier) [inline config/settings.py]: could be `DEBUG = os.getenv('DEBUG') == True`

File: config/settings.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+ALLOWED_HOSTS = [
```
Comment (richtier) [inline config/settings.py]:
```
# As the app is running behind a host-based router supplied by Heroku or other
# PaaS, we can open ALLOWED_HOSTS
ALLOWED_HOSTS = ['*']
```

File: config/settings.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+WSGI_APPLICATION = 'config.wsgi.application'
+
+
+# Database
+# https://docs.djangoproject.com/en/1.11/ref/settings/#databases
+
+DATABASES = {
+    'default': {
+        'ENGINE': 'django.db.backends.sqlite3',
+        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
```
Comment (richtier) [inline config/settings.py]: we use `DATABSE_URL` and django-database-url in other apps. it's convenient, and consistency simplifies maintenance.

File: config/urls.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+Function views
+    1. Add an import:  from my_app import views
+    2. Add a URL to urlpatterns:  url(r'^$', views.home, name='home')
+Class-based views
+    1. Add an import:  from other_app.views import Home
+    2. Add a URL to urlpatterns:  url(r'^$', Home.as_view(), name='home')
+Including another URLconf
+    1. Import the include() function: from django.conf.urls import url, include
+    2. Add a URL to urlpatterns:  url(r'^blog/', include('blog.urls'))
+"""
```
Comment (richtier) [inline config/urls.py]: these comments can be removed. they're available in django docs

File: config/settings.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+STATIC_ROOT = os.path.join(BASE_DIR, 'staticfiles')
+STATIC_HOST = os.environ.get('STATIC_HOST', '')
+STATIC_URL = STATIC_HOST + '/static/'
+STATICFILES_STORAGE = 'whitenoise.storage.CompressedManifestStaticFilesStorage'
+
+
+PIR_API_KEY = os.getenv('PIR_API_KEY')
+PIR_API_URL = os.getenv('PIR_API_URL')
+
```
Comment (richtier) [inline config/settings.py]: please add sentry integration so unhandled exceptions are logged to sentry (like all other of our services do).

File: pir_frontend/forms.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
+from django.conf import settings
+
+from directory_components import forms, fields
+from django_countries.data import COUNTRIES
+from nocaptcha_recaptcha.fields import NoReCaptchaField
+
+from pir_client import PIRAPIClient
+
+
+class PIRForm(forms.Form):
```
Comment (richtier) [inline pir_frontend/forms.py]: the name of this form does not help the reader understand what it is/does/

please rename to something more specific.

File: pir_frontend/forms.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    email = fields.EmailField(
+        required=True,
+        label='Email',
+    )
+
+    country = fields.ChoiceField(
+        required=True,
+        label='Country',
+        choices=sorted(
+            [(k, v) for k, v in COUNTRIES.items()],
```
Comment (richtier) [inline pir_frontend/forms.py]: it looks like `(k, v) for k, v in COUNTRIES.items()` is the same as `COUNTRIES.items()`

File: pir_frontend/forms.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+                o['value'],
+                o['display_name']) for o in options['sector']['choices']
+        ]
+
+        self.fields['sector'] = fields.ChoiceField(
+            label='Sector',
+            choices=sector_choices
+        )
+
+        self.fields['captcha'] = NoReCaptchaField()
```
Comment (richtier) [inline pir_frontend/forms.py]: why adding this in `__init__`?

File: pir_frontend/forms.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+        sector_choices = [
+            (
+                o['value'],
+                o['display_name']) for o in options['sector']['choices']
+        ]
+
+        self.fields['sector'] = fields.ChoiceField(
+            label='Sector',
+            choices=sector_choices
+        )
```
Comment (richtier) [inline pir_frontend/forms.py]: would be better for the reader if the field was defined at class level and the choices set in `__init__`

File: pir_frontend/tests/test_views.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+            },
+        ]
+    }
+}
+
+
+class ViewTest(TestCase):
+
+    @patch('pir_frontend.forms.PIRAPIClient')
+    def test_pir_view(self, client_instance_mock):
```
Comment (richtier) [inline pir_frontend/tests/test_views.py]: please restructure this to be "one thing tested per test"

also, the name should help the reader explain what the test does.

File: pir_frontend/views.py (hunk 7 above)
Comment (richtier) [reply pir_frontend/views.py]: also the method `handle_exception` is a good place to handle this logic

File: pir_frontend/views.py [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+                    request, 'index.html', {
+                        'error': (
+                            'Something is wrong with the service.'
+                            ' Please try again later'
+                        )
+                    },
+                    status=500
+                )
+
+            return render(
```
Comment (richtier) [inline pir_frontend/views.py]: returning a `TemplateResponse` makes the view easier to test: with a `TemplateResponse` we can test what template name was used.

File: pir_frontend/views.py [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+                status=201
+            )
+        else:
+            return render(
+                request, 'index.html', {'form': form},
+                status=400
+            )
+
+    def get(self, request):
+        return render(
```
Comment (richtier) [inline pir_frontend/views.py]: looks like this is a good candidate for a `FormView`

File: pir_frontend/views.py (hunk 7 above)
Comment (sdonk) [reply pir_frontend/views.py]: This should go to sentry, add it

File: config/settings.py (hunk 11 above)
Comment (rollokb) [reply config/settings.py]: This isn't using a database.

File: config/settings.py (hunk 9 above)
Comment (rollokb) [reply config/settings.py]: It could be.

File: pir_frontend/forms.py (hunk 14 above)
Comment (rollokb) [reply pir_frontend/forms.py]: Personalised Investment Report Form. I'll rename it to that.

File: pir_frontend/forms.py (hunk 17 above)
Comment (rollokb) [reply pir_frontend/forms.py]: The choices are dynamically constructed. Depending on a result each time the form is constructed. 

I did just check, and I think you can make choices a callable now. I'll see if that works.

**Pull request merged**